File: src/2958-Length-of-Longest-Subarray-With-at-Most-K-Frequency/length_of_longest_subarray_with_at_most_k_frequency.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <algorithm>

class Solution {
public:
    int maxSubarrayLength(std::vector<int>& nums, int k) {
        int left = 0;
        int right = 0;
        int ans = 0;
        std::unordered_map<int, int> frequency;
        int n = nums.size();

        for (right = 0; right < n; right++) {
            if (frequency.find(nums[right]) == frequency.end()) {
                frequency[nums[right]] = 0;
            }
            frequency[nums[right]] += 1;
            while (frequency[nums[right]] > k) {
                frequency[nums[left]] -= 1;
                left += 1;
            }
            ans = std::max(ans, right - left + 1);
        }
        
        return ans;
    }
};

int main() {
    std::vector<int> nums = {1,2,3,1,2,3,1,2};
    int k = 2;
    Solution solution;
    
    std::cout << solution.maxSubarrayLength(nums, k);

    return 0;
}

File: src/2958-Length-of-Longest-Subarray-With-at-Most-K-Frequency/length_of_longest_subarray_with_at_most_k_frequency.py
from typing import List, Dict

class Solution:
    def maxSubarrayLength(self, nums: List[int], k: int) -> int:
        left: int = 0
        right: int = 0
        ans: int = 0
        frequency: Dict[int, int] = {}
        n: int = len(nums)

        for right in range(n):
            if nums[right] not in frequency:
                frequency[nums[right]] = 0  
            frequency[nums[right]] += 1
            while frequency[nums[right]] > k:
                frequency[nums[left]] -= 1
                left += 1
            
            ans = max(ans, right - left + 1)
        
        return ans


if __name__ == "__main__":
    nums: List[int] = [1,2,3,1,2,3,1,2]
    k: int = 2
    solution = Solution()
    
    print(solution.maxSubarrayLength(nums, k))